### PR TITLE
Add OAuth with account level app

### PR DESCRIPTION
**Pull request by larkox**
#### Summary
Right now, Zoom OAuth allow to use User managed apps for OAuth. This means every user has to sign in to his account. We are adding an Account level app capability. This means that only the sys admin connects to the Zoom account, and then every user can connect based on their e-mail, like with Zoom JWT.

#### Ticket Link
None

#### Waiting for
Version v1.4.0 released

### CODE REVIEW

File: server/plugin.go
Comment (mickmister) [inline server/plugin.go:278]: 0/5 to use `p.getSiteURL()` here

File: server/command.go
Comment (larkox) [reply server/command.go:63]: No strong opinion on this. My idea was to not fill the user with unneeded information, like knowing that there is a connect option, or that they are in certain type of configuration. Nevertheless, we can put something like:
"You do not need to connect to use the plugins"

File: server/command.go
Comment (larkox) [reply server/command.go:120]: With the current architecture, we are never returning errors. If I am not mistaken, the reason is that the UX is way worse.

File: server/plugin.go
Comment (iomodo) [inline server/plugin.go:381]: `getZoomUserWithSuperUserToken` and `getZoomUserWithToken` is basically the same code, can't we remove some duplicates?

File: server/plugin.go
Comment (iomodo) [inline server/plugin.go:279]: Shouldn't we pass `false` here for more clarity?

File: server/http.go
Comment (iomodo) [inline server/http.go:468]: Why not `*p.API.GetConfig().ServiceSettings.SiteURL, channelID, "false")` ?

File: server/http.go
Comment (iomodo) [inline server/http.go:171]: Why aren't we posting meeting for a `AccountLevelApp`?

File: server/command.go
Comment (larkox) [reply server/command.go:62]: Connect is hidden if:
- We are using JWT (there is no need to connect)
- We are using Account Level OAuth and you are not a sys admin (only sysadmins can connect)

Connect is visible if:
- We are using User Level OAuth (every user has to connect)
- We are using Account Level Oauth and you are a sys admin (sys admins connect account level apps)

Or at least that is how I tried to implement it.

File: server/http.go
Comment (iomodo) [inline server/http.go:161]: We could do it outside `if` to avoid duplication.

File: server/http.go
Comment (iomodo) [inline server/http.go:188]: shouldn't we check for the error here?

**Comment by larkox** [comment]
Heads up to @mickmister and @iomodo that during conflict resolution on merge, a few things were moved and refactored, in case you want to give it a second look as Dev Review.

**Comment by DHaussermann** [comment]
@larkox now that we have the related PR merged, can you please look at these conflicts ☝️  so we can move this forward into testing?

**Comment by larkox** [comment]
@DHaussermann I have solved the merge conflicts, and it should be ready to go. I haven't give the time to do a proper testing on my side, so there might be many issues around. If you have the time, I would love for you to test it. If not, I will give it a proper round of testing soon.

**Comment by larkox** [comment]
I re-requested @mickmister and @iomodo review since the changes in the code after the merge are big enough to add another review.

**Comment by DHaussermann** [comment]
@larkox I saw another issue here. On the last round of dev. It seems like when you enable and disable the feature a couple times the sate can get broken. This was a temporary issue. It stopped occurring after a few minutes. It was very strange as it stopped happening without either the user or the admin refreshing their session. 

When I toggle the feature on and off now I I see Invalid oAuth state for the admin and `Error fetching user: 401 error returned while fetching zoom user` for the user. 

Is it possible this is cause because the admin and user are sharing the same set of oAuth credentials in different sessions because it's a dev account? 

What I was seeing:
- `OAuth by Account Level App (Beta)` is true
- When a user tries to connect they get  `Error fetching user: error fetching zoom user: Get "https://api.zoom.us/v2/users/dylan.haussermann@gmail.com": oauth2: cannot fetch token: 401 Unauthorized
Response: {"reason":"Invalid Token!","error":"invalid_request"}` regardless of if email is valid. In this case the admin has not connected since the feature was enabled so they should be getting the post to contact the sys admin
- when a sysadmin tries to connect they get `invalid OAuth user state` in the window that open when clicking the connection link

Steps (as close as I can recall) : 
- Install Zoom and enable the new feature
- Connect as an admin
- Launch some meetings successfully as user with a valid email
- Disable the feature
- Attempt to launch a meeting as a user and note they must connect individually - as expected
- Map the user and launch a meeting 
- After the meeting has ended. login as the admin and re-enable the new feature

**Comment by DHaussermann** [comment]
@aaronrothschild the feature is label in the config as `OAuth by Account Level App (Beta)` . Is this standard terminology you or @larkox  have taken from somewhere? If I pretend I have never read this before and look at it, I'm not sure if this is app level or account level oAuth 🤷‍♂️  

The description is nice and clear. If this is not a common term, we should consider clarifying it.

**Comment by DHaussermann** [comment]
Deleted previous comment about the label. After reviewing the terminology Zoom uses for "_Account-level app_" our config label is consistent. Please disregard.

File: server/command.go
Comment (mickmister) [inline server/command.go:137]: Can we use a `GetSiteURL` method here? Looks like we use `p.siteURL` elsewhere too. Maybe just use that

**Review by mickmister — APPROVED** [review]
Great work!! One non-blocking request to avoid dereferencing `siteURL` without checking for non-nil first

**Review by iomodo — APPROVED** [review]
LGTM!

**Comment by larkox** [comment]
@DHaussermann In theory, if you change between different methods, you should provide different app ids and secrets, since one is for account level, and the other one is for user level. It make sens that even if it allow you to connect, you receive 401 (unauthorized) when trying to create a meeting (the token stored does not provide the needed permissions).

Does this make sense?

**Comment by DHaussermann** [comment]
Thanks for pointing this out @larkox Given that the account type would also change this seems like an edge case as is.

**Comment by DHaussermann** [comment]
Tested and passed
- Mapping as admin who does not have a valid Zoom email
- Mapping as admin who has a valid Zoom email

- User tries to launch Zoom via icon and slash command when admin has not connected
- User successfully launches meetings via icon and slash command when user is connected
- User tries to launch Zoom via icon and slash command when admin has not connected 

- Attempt to map as a non-admin `Zoom App not connected. Contact your System administrator.` No connection link
Switch back and forth between Account level oAuth and Application level oAuth
- Add OAuth with account level app feature is always set to False on fresh install even if Oauth is true.

LGHM!

**Review by DHaussermann — APPROVED** [review]
See testing overview ☝️

**Pull request merged**
